core/dmux.go

```diff
@@ -311,10 +358,204 @@ func simpleSetup(size, qsize int, sink Sink) ([]chan interface{}, *sync.WaitGrou
+				if sidelineMetaByteArrayErr != nil {
+					//TODO : think about this
+					log.Printf("error in serde of SidelineMeta")
+					return errors.New("error in serde of SidelineMeta")
+				}
+				val := source.GetValue(channelObject.Msg)
+				key := source.GetKey(channelObject.Msg)
+				partition := source.GetPartition(channelObject.Msg)
+				offset := source.GetOffset(channelObject.Msg)
+				log.Printf("Inside sideline channel for partition %d offset %d", partition, offset)
+				kafkaSidelineMessage := sideline_models.SidelineMessage{
+					GroupId:           string(key),
+					Partition:         partition,
+					EntityId:          string(key) + sideline.ConsumerGroupName + sideline.ClusterName,
+					Offset:            offset,
+					ConsumerGroupName: sideline.ConsumerGroupName,
+					ClusterName:       sideline.ClusterName,
+					Message:           val,
+					Version:           channelObject.Version,
+					ConnectionType:    sideline.ConnectionType,
+					SidelineMeta:      sidelineMetaByteArray,
+				}
+
+				sidelineByteArray, err := json.Marshal(kafkaSidelineMessage)
+				if err != nil {
+					log.Printf("error in serde of kafkaSidelineMessage " + err.Error())
+					return errors.New("error in serde of kafkaSidelineMessage " + err.Error())
+				}
+				sidelineMessageResponse := sidelineImpl.SidelineMessage(sidelineByteArray)
+				if !sidelineMessageResponse.Success {
+					var check sideline_models.CheckMessageSidelineResponse
+					log.Printf(sidelineMessageResponse.ErrorMessage)
+					if sidelineMessageResponse.ConcurrentModificationError != nil {
+						checkSidelineMessage := sideline_models.SidelineMessage{
+							GroupId:           string(key),
+							Partition:         partition,
+							EntityId:          string(key) + sideline.ConsumerGroupName + sideline.ClusterName,
+							Offset:            offset,
+							ConsumerGroupName: sideline.ConsumerGroupName,
+							ClusterName:       sideline.ClusterName,
+							Message:           val,
+							ConnectionType:    sideline.ConnectionType,
+						}
+						checkSidelineMessageBytes, checkSidelineMessageErr := json.Marshal(checkSidelineMessage)
+						if checkSidelineMessageErr != nil {
+							log.Printf("error in serde of checkSidelineMessage")
+							return errors.New("error in serde of checkSidelineMessage")
+						}
+						checkBytes, checkErr := sidelineImpl.(sideline_models.CheckMessageSideline).
+							CheckMessageSideline(checkSidelineMessageBytes)
+						err := json.Unmarshal(checkBytes, &check)
+						if err != nil {
+							log.Printf("error in serde of CheckMessageSidelineResponse")
+							return errors.New("error in serde of CheckMessageSidelineResponse")
+
+						}
+						if checkErr != nil {
+							log.Printf("Error in checking if message is sidelined " + checkErr.Error())
+							return errors.New("Error in checking if message is sidelined " + checkErr.Error())
+						}
+						channelObject.Version = check.Version
+						return errors.New("Retrying ")
+					}
```

Comment:
Isn't this piece internal to how sideline works? shouldn't sideline implementation handle this and emit final reason for why message cannot be sidelined?

Reply:
the idea was to provide sideline as a capability from dmux 
hence we had created models for the communication between client and dmux 
if the contracts are generic and if any new client can honour them then the above can be viewed as a genric implementation.